frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java

```diff
@@ -0,0 +1,91 @@
+import static com.virtuslab.gitmachete.frontend.resourcebundles.GitMacheteBundle.getString;
+
+import java.nio.file.Path;
+
+import com.intellij.openapi.progress.ProgressIndicator;
+import com.intellij.openapi.progress.Task;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.vcs.VcsNotifier;
+import git4idea.repo.GitRepository;
+import lombok.experimental.ExtensionMethod;
+import lombok.val;
+import org.checkerframework.checker.guieffect.qual.UIEffect;
+
+import com.virtuslab.binding.RuntimeBinding;
+import com.virtuslab.branchlayout.api.BranchLayout;
+import com.virtuslab.branchlayout.api.readwrite.IBranchLayoutWriter;
+import com.virtuslab.gitmachete.frontend.file.MacheteFileWriter;
+import com.virtuslab.gitmachete.frontend.vfsutils.GitVfsUtils;
+import com.virtuslab.qual.guieffect.UIThreadUnsafe;
+
+@ExtensionMethod(GitVfsUtils.class)
+public class RenameBackgroundable extends Task.Backgroundable {
+
+  protected final Project project;
+  protected final GitRepository gitRepository;
+  protected final BranchLayout branchLayout;
+  protected final Runnable renameRunnable;
+  protected final String currentBranchName;
+  protected final String newBranchName;
+
+  public RenameBackgroundable(
+      GitRepository gitRepository,
+      BranchLayout branchLayout,
+      Runnable renameRunnable,
+      String currentBranchName,
+      String newBranchName) {
+    super(gitRepository.getProject(), getString("action.GitMachete.RenameBackgroundable.task-title"));
+    this.project = gitRepository.getProject();
+    this.gitRepository = gitRepository;
+    this.branchLayout = branchLayout;
+    this.renameRunnable = renameRunnable;
+    this.currentBranchName = currentBranchName;
+    this.newBranchName = newBranchName;
+  }
+
+  @Override
+  @UIThreadUnsafe
+  public void run(ProgressIndicator indicator) {
+    Path macheteFilePath = gitRepository.getMacheteFilePath();
+
+    val newBranchLayout = branchLayout.rename(currentBranchName, newBranchName);
+    val branchLayoutWriter = RuntimeBinding.instantiateSoleImplementingClass(IBranchLayoutWriter.class);
+
+    runWriteActionOnUIThread(() -> {
+      MacheteFileWriter.writeBranchLayout(
+          macheteFilePath,
+          branchLayoutWriter,
+          newBranchLayout,
+          /* backupOldLayout */ true,
+          /* requestor */ this);
+
+      renameRunnable.run();
+
+      // `renameRunnable` may perform some sneakily-asynchronous operations (e.g. renameBranch).
+      // The high-level method used within the runnable does not allow us to schedule the tasks after them.
+      // (Stepping deeper is not an option since we would lose some important logic or become very dependent on the internals of git4idea).
+      // Hence, we wait for the creation of the branch (with exponential backoff).
+      //
+      // Theoretically we could not wait for the branch creation.
+      // However, in would result in a premature execution of GitMacheteRepositoryUpdateBackgroundable,
```

Comment:
As commented elsewhere... this reasoning is questionable 🤔

Reply:
If we check empirically that it works as expected... let's stick to that. But still, the comment will need to be updated

Reply:
anyway we need some specific comment here